File: 05-Two-Pointers/valid_palindrome.cpp
/*
Problem Name: Valid Palindrome
Platform: LeetCode
Difficulty: Easy

Approach 1: Reverse string and compare. Uses O(n) space.
Approach 2: Two pointers moving inwards comparing valid alphanumeric characters.
Optimal Approach: Two pointers moving inwards in-place.

Time Complexity: O(n)
Space Complexity: O(1)

Pattern: Two Pointers
*/

#include <iostream>
#include <string>
#include <cctype>
#include <cassert>

class Solution {
public:
    bool isPalindrome(std::string s) {
        int left = 0;
        int right = s.length() - 1;
        
        while (left < right) {
            while (left < right && !std::isalnum(s[left])) {
                left++;
            }
            while (left < right && !std::isalnum(s[right])) {
                right--;
            }
            
            if (std::tolower(s[left]) != std::tolower(s[right])) {
                return false;
            }
            
            left++;
            right--;
        }
        
        return true;
    }
};

int main() {
    Solution sol;
    assert(sol.isPalindrome("A man, a plan, a canal: Panama") == true);
    assert(sol.isPalindrome("race a car") == false);
    std::cout << "Valid Palindrome C++: All tests passed! 🎉" << std::endl;
    return 0;
}

File: 05-Two-Pointers/valid_palindrome.py
# """
# Problem Name: Valid Palindrome
# Platform: LeetCode
# Difficulty: Easy
#
# Approach 1: Filter alphanumeric characters, convert to lowercase, reverse string and check equality. O(n) space.
# Approach 2: Two pointers moving inwards comparing lowercase alphanumeric characters in-place.
# Optimal Approach: Two pointers in-place.
#
# Time Complexity: O(n)
# Space Complexity: O(1)
#
# Pattern: Two Pointers
# """

class Solution:
    def isPalindrome(self, s: str) -> bool:
        left, right = 0, len(s) - 1
        
        while left < right:
            while left < right and not s[left].isalnum():
                left += 1
            while left < right and not s[right].isalnum():
                right -= 1
                
            if s[left].lower() != s[right].lower():
                return False
                
            left += 1
            right -= 1
            
        return True

if __name__ == "__main__":
    sol = Solution()
    assert sol.isPalindrome("A man, a plan, a canal: Panama") is True
    assert sol.isPalindrome("race a car") is False
    print("Valid Palindrome Python: All tests passed! 🎉")
